### PR TITLE
Burn train api refactor

### PR DESCRIPTION
**Pull request by Charles23R**
### Checklist

- [ ] Confirmed that `cargo run-checks` command has been executed.
- [ ] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

_Provide links to relevant issues and dependent PRs._

### Changes

_Summarize the problem being addressed and your solution._

### Testing

_Describe how these changes have been tested._

### CODE REVIEW

File: crates/burn-train/src/learner/supervised/paradigm.rs [hunk 1]
```diff
@@ -149,9 +173,13 @@ where
...
 
     /// Update the checkpointing_strategy.
-    pub fn with_checkpointing_strategy<CS>(mut self, strategy: CS) -> Self
+    pub fn with_checkpointing_strategy(
+        mut self,
+        strategy: <SC::PC as ParadigmComponentsTypes>::CheckpointerStrategy,
+    ) -> Self
     where
-        CS: CheckpointingStrategy + 'static,
+        // CS: CheckpointingStrategy + 'static,
```
Comment (nathanielsimard) [inline crates/burn-train/src/learner/supervised/paradigm.rs]: deadcode

File: crates/burn-train/src/learner/supervised/paradigm.rs [hunk 2]
```diff
@@ -366,133 +401,104 @@ where
...
-        if let LearningStrategy::MultiDevice(devices, _) = &learning_strategy
-            && devices.len() == 1
-        {
-            return LearningStrategy::SingleDevice(devices[0].clone());
+        match self.training_strategy {
+            TrainingStrategy::SingleDevice(device) => {
+                let single_device: SingleDevicetrainingStrategy<SC> =
+                    SingleDevicetrainingStrategy::new(device);
+                single_device.train(
+                    // self.learner,
```
Comment (nathanielsimard) [inline crates/burn-train/src/learner/supervised/paradigm.rs]: deadcode

File: crates/burn-train/src/metric/processor/minimal.rs [hunk 3]
```diff
@@ -9,6 +9,7 @@ use std::sync::Arc;
 ///   - Computing and storing metrics in an [event store](crate::metric::store::EventStore).
 #[allow(dead_code)]
 #[derive(new)]
+// TODO : DELETE
```
Comment (nathanielsimard) [inline crates/burn-train/src/metric/processor/minimal.rs]: Still valid?

File: crates/burn-train/src/learner/supervised/paradigm.rs [hunk 4]
```diff
@@ -17,90 +11,107 @@ use crate::metric::processor::{
...
-    M::InnerModule: ValidStep<VI, VO>,
-    O: Optimizer<M, B>,
-    S: LrScheduler,
-    TI: Send + 'static,
-    VI: Send + 'static,
-    TO: ItemLazy + 'static,
-    VO: ItemLazy + 'static,
-{
+/// Structure to configure and launch supervised learning trainings.
+pub struct SupervisedTraining<SC: SupervisedLearningComponentsTypes> {
```
Comment (laggui) [inline crates/burn-train/src/learner/supervised/paradigm.rs:36]: I'm not sure if `SupervisedTraining` might be too restrictive, given that it might be possible to leverage the same training loop even in other contexts (e.g., self-supervised or unsupervised which still have similar training loop, just the signal is different). I know they all mention "supervised" but people generally make the distinction :smile: 

But we have time to think about this before introducing this breaking change in a major release.

Otherwise, LGTM!!
